projects/mci/src/mci_broadcast.c

```diff
@@ -24,7 +49,105 @@ static void prv_broadcast_bus_measurement(MotorControllerBroadcastStorage *stora
                                    (uint16_t)measurements[RIGHT_MOTOR_CONTROLLER].bus_current_a);
 }
 
+static void prv_broadcast_status(MotorControllerBroadcastStorage *storage) {
+  uint32_t *measurements = storage->measurements.status;
+  CAN_TRANSMIT_MOTOR_STATUS(measurements[LEFT_MOTOR_CONTROLLER],
+                            measurements[RIGHT_MOTOR_CONTROLLER]);
+}
+
+// Change the MCP2515 filter to filter for the next ID to look for
+static void prv_change_filter(MotorControllerBroadcastStorage *storage) {
+  if (storage->cb_storage.cur_measurement == NUM_MOTOR_CONTROLLER_BROADCAST_MEASUREMENTS - 1) {
+    storage->cb_storage.cur_measurement = MOTOR_CONTROLLER_BROADCAST_STATUS;
+#ifndef RIGHT_MOTOR_CONTROLLER_UNUSED
+    storage->cb_storage.motor_controller = !storage->cb_storage.motor_controller;
+#endif
+  } else {
+    storage->cb_storage.cur_measurement++;
+  }
+  uint32_t filter = s_base_addr_lookup[storage->cb_storage.motor_controller] +
+                    s_offset_lookup[storage->cb_storage.cur_measurement];
+  // MCP2515 requires both filters to be set, so just use the same one twice
+  // Looking for multiple message IDs seems to cause issues, so we iterate through all IDs required
+  // one by one
+  LOG_DEBUG("Now filtering for ID 0x%x\n", (int)filter);
+  uint32_t filters[2] = { filter, filter };
+  mcp2515_set_filter(storage->motor_can, filters);
+}
+
+// CB for rx received
+static void prv_process_rx(uint32_t id, bool extended, uint64_t data, size_t dlc, void *context) {
+  MotorControllerBroadcastStorage *storage = context;
+  LOG_DEBUG("Received rx from id: 0x%x\n", (int)id);
+  LOG_DEBUG("Data: 0x%" PRIx64 "\n", data);
+  // calculate the base offset
+  uint32_t cur_mc_id = s_base_addr_lookup[storage->cb_storage.motor_controller];
+  uint32_t offset = (id - cur_mc_id);
+  // map base offset to the cb array index
+  uint32_t cb_index = NUM_MOTOR_CONTROLLER_BROADCAST_MEASUREMENTS;
+  for (uint32_t i = 0; i < NUM_MOTOR_CONTROLLER_BROADCAST_MEASUREMENTS; i++) {
+    if (s_offset_lookup[i] == offset) {
+      cb_index = i;
+      break;
+    }
+  }
+  // check if received message ID doesn't have a CB index
+  if (cb_index == NUM_MOTOR_CONTROLLER_BROADCAST_MEASUREMENTS) {
+    LOG_WARN("WARNING - no cb for message ID 0x%x or received message from wrong MC\n", (int)id);
+    // TODO(SOFT-139): error handling here
+    // This should NEVER happen, so it may make sense to throw an error
+    // However, have had this happen a few times, always with offset 9 (voltage rail measurement)
+    // not sure what the cause is, doesn't happen often
```

Comment:
Yeah, it's really bizarre, there's no reason why it should be happening since we don't (intentionally) filter for it at any point.  I assume it's driver weirdness, but IIRC I've only ever experienced it when testing with the MCP2515 currently in MCI so I guess there's an extremely small chance that it's a very, very weird hardware issue.

Reply:
Idea: could it be a critical section issue? I.e. while we're changing the filter, could there be a brief intermediate state where the voltage rail message is allowed? If so, unlucky soft timer / interrupt timing could result in rxing a message in that time. If this is the case, it might be able to be fixed by guarding any critical sections (which shouldn't be interrupted so as not to be in an invalid state during an interrupt), but idk how MCP2515 works. That's the only way I can think it could be a firmware issue anyways.

Reply:
Whoops missed this reply until now, sorry 😐 

I believe that that isn't possible since, once we put the MCP2515 into config mode, I _think_ it doesn't listen to the CAN bus until after we put it back into normal mode.  These changes happen at the start/end of the `mcp2515_set_filter` function, and all the register-level filter changes happen with it in config mode. The datasheet isn't 100% clear on this, but it explicitly specifies on page 59 when other nodes do listen to the bus (and it'd be really weird to allow RXing messages while configuring the masks/filters).

That said, I'll keep this in mind for the next time I get access to the hardware as something to double-check to confirm.